### PR TITLE
coq: Move dune, ocaml and findlib to nativeBuildInputs

### PR DESCRIPTION
**Pull request by ulrikstrid**
###### Motivation for this change

When overriding with `csdp = null;` coq fails to build (at least when cross compiling).
Moving dune to nativeBuildInputs and adding ocaml, findlib and menhir to nativeBuildInputs fixes this issue.
Also moving the `coqide` behind the `buildIde` flag.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Zimmi48 — COMMENTED** [review]
I've never really understood cross-compilation, so I cannot comment on the `nativeBuildInputs` / `buildInputs` difference (if it solves the issue, then good), but:

1. Shouldn't all the `ocamlBuildInputs` be included in `nativeBuildInputs` then (instead of some being duplicated)?
2. Why the menhir dependency? AFAICT, Coq does not depend on menhir and never has.

**Comment by ulrikstrid** [comment]
> Shouldn't all the ocamlBuildInputs be included in nativeBuildInputs then (instead of some being duplicated)?

I changed it so some of the `ocamlBuildInputs` are now in a new `ocamlNativeBuildInputs` so that they are split correctly, I missed this part in the first commit.

> Why the menhir dependency? AFAICT, Coq does not depend on menhir and never has.

I thought I saw menhir somewhere, it is now removed, thanks for pointing it out

Could you have another look @Zimmi48? Thanks!

**Review by Zimmi48 — APPROVED** [review]

**Comment by vbgl** [comment]
@GrahamcOfBorg build coqPackages.equations

**Review by vbgl — CHANGES_REQUESTED** [review]
This seems to break many things.

**Comment by Zimmi48** [comment]
If a new `extraNativeBuildInputs` optional argument is added to `mkDerivation`, then the documentation in `doc/languages-frameworks/coq.section.md` should be updated, and some guidance should be provided as to when use `extraNativeBuildInputs` vs `extraBuildInputs`.

**Comment by vbgl** [comment]
Marking as draft for now. Please ping again when ready.

**Comment by ulrikstrid** [comment]
This seems to be in a working state now. I updated the documentation about when to use what as well.

**Comment by ulrikstrid** [comment]
@vbgl would you be able to do a new review? I can post the nixpkgs-review result in a bit as well

**Comment by ulrikstrid** [comment]
nixpkgs-review output:

<details>
<summary>93 packages updated:</summary>
coq coq coq coq coq coq coq coq coq coq coq coq coq coq8.13-aac-tactics coq8.13-addition-chains coq8.13-autosubst coq8.13-bignums coq8.13-category-theory coq8.13-ceres coq8.13-cheerios coq8.13-color coq8.13-compcert coq8.13-compcert coq8.13-coq-bits coq8.13-coq-ext-lib coq8.13-coq-record-update coq8.13-CoqEAL coq8.13-coqhammer coq8.13-coqprime coq8.13-coqtail-math coq8.13-coquelicot coq8.13-corn coq8.13-deriving coq8.13-dpdgraph coq8.13-elpi coq8.13-equations coq8.13-extructures coq8.13-flocq coq8.13-fourcolor coq8.13-gaia coq8.13-gaia-hydras coq8.13-gappalib coq8.13-goedel coq8.13-graph-theory coq8.13-hierarchy-builder coq8.13-hydra-battles coq8.13-InfSeqExt coq8.13-InteractionTrees coq8.13-interval coq8.13-iris coq8.13-itauto coq8.13-LibHyps coq8.13-math-classes coq8.13-mathcomp-algebra coq8.13-mathcomp-all coq8.13-mathcomp-character coq8.13-mathcomp-field coq8.13-mathcomp-fingroup coq8.13-mathcomp-solvable coq8.13-mathcomp-ssreflect coq8.13-mathcomp1.14-abel coq8.13-mathcomp1.14-analysis coq8.13-mathcomp1.14-bigenough coq8.13-mathcomp1.14-finmap coq8.13-mathcomp1.14-multinomials coq8.13-mathcomp1.14-real-closed coq8.13-mathcomp1.14-tarjan coq8.13-mathcomp1.14-word coq8.13-mathcomp1.14-zify coq8.13-metalib coq8.13-odd-order coq8.13-paco coq8.13-paramcoq coq8.13-parsec coq8.13-pocklington coq8.13-QuickChick coq8.13-reglang coq8.13-relation-algebra coq8.13-semantics coq8.13-serapi coq8.13-simple-io coq8.13-smpl coq8.13-stdpp coq8.13-StructTact coq8.13-tlc coq8.13-topology coq8.13-verdi coq8.13-VST coq8.13-zorns-lemma frama-c ocaml4.12.0-ligo satallax why3
</details>
<details>
<summary>91 packages built:</summary>
compcert coq coqPackages.Cheerios coqPackages.CoLoR coqPackages.ITree coqPackages.InfSeqExt coqPackages.LibHyps coqPackages.QuickChick coqPackages.StructTact coqPackages.VST coqPackages.Verdi coqPackages.aac-tactics coqPackages.addition-chains coqPackages.autosubst coqPackages.bignums coqPackages.category-theory coqPackages.ceres coqPackages.compcert coqPackages.coq-bits coqPackages.coq-elpi coqPackages.coq-ext-lib coqPackages.coq-record-update coqPackages.coqeal coqPackages.coqhammer coqPackages.coqprime coqPackages.coqtail-math coqPackages.coquelicot coqPackages.corn coqPackages.deriving coqPackages.dpdgraph coqPackages.equations coqPackages.extructures coqPackages.flocq coqPackages.fourcolor coqPackages.gaia coqPackages.gaia-hydras coqPackages.gappalib coqPackages.goedel coqPackages.graph-theory coqPackages.hierarchy-builder coqPackages.hydra-battles coqPackages.interval coqPackages.iris coqPackages.itauto coqPackages.math-classes coqPackages.mathcomp coqPackages.mathcomp-abel coqPackages.mathcomp-algebra coqPackages.mathcomp-analysis coqPackages.mathcomp-bigenough coqPackages.mathcomp-character coqPackages.mathcomp-field coqPackages.mathcomp-fingroup coqPackages.mathcomp-finmap coqPackages.mathcomp-real-closed coqPackages.mathcomp-solvable coqPackages.mathcomp-ssreflect coqPackages.mathcomp-tarjan coqPackages.mathcomp-word coqPackages.mathcomp-zify coqPackages.metalib coqPackages.multinomials coqPackages.odd-order coqPackages.paco coqPackages.paramcoq coqPackages.parsec coqPackages.pocklington coqPackages.reglang coqPackages.relation-algebra coqPackages.semantics coqPackages.serapi coqPackages.simple-io coqPackages.smpl coqPackages.stdpp coqPackages.tlc coqPackages.topology coqPackages.zorns-lemma coq_8_10 coq_8_11 coq_8_12 coq_8_14 coq_8_15 coq_8_5 coq_8_6 coq_8_7 coq_8_8 coq_8_9 framac ligo satallax why3</details>

Not sure why there is a difference in the packages built since there are not failures...

File: doc/languages-frameworks/coq.section.md [hunk 1]
```diff
@@ -29,7 +29,8 @@ The recommended way of defining a derivation for a Coq library, is to use the `c
 * `releaseRev` (optional, defaults to `(v: v)`), provides a default mapping from release names to revision hashes/branch names/tags,
 * `displayVersion` (optional), provides a way to alter the computation of `name` from `pname`, by explaining how to display version numbers,
 * `namePrefix` (optional, defaults to `[ "coq" ]`), provides a way to alter the computation of `name` from `pname`, by explaining which dependencies must occur in `name`,
-* `extraBuildInputs` (optional), by default `buildInputs` just contains `coq`, this allows to add more build inputs,
+* `extraNativeBuildInputs` (optional), by default `nativeBuildInputs` just contains `coq`, this allows to add more native build inputs, nativeBuildInputs are executalbes and buildInputs are libraries and dependencies,
```
Comment (vbgl) [inline doc/languages-frameworks/coq.section.md:32]:
```suggestion
* `extraNativeBuildInputs` (optional), by default `nativeBuildInputs` just contains `coq`, this allows to add more native build inputs, `nativeBuildInputs` are executables and `buildInputs` are libraries and dependencies,
```

File: pkgs/development/coq-modules/coq-bits/default.nix [hunk 2]
```diff
@@ -22,7 +22,7 @@ with lib; mkCoqDerivation {
     };
   };
 
-  propagatedBuildInputs = [ mathcomp.algebra ];
+  propagatedBuildInputs = [ mathcomp.algebra mathcomp.ssreflect mathcomp.fingroup ];
```
Comment (vbgl) [inline pkgs/development/coq-modules/coq-bits/default.nix:25]: Is this change related to the purpose of this PR?

File: pkgs/development/coq-modules/coquelicot/default.nix [hunk 3]
```diff
@@ -16,8 +16,8 @@ with lib; mkCoqDerivation {
   release."3.0.2".sha256 = "1rqfbbskgz7b1bcpva8wh3v3456sq2364y804f94sc8y5sij23nl";
   releaseRev = v: "coquelicot-${v}";
 
-  nativeBuildInputs = [ which autoconf ];
-  propagatedBuildInputs = [ ssreflect ];
+  extraNativeBuildInputs = [ which autoconf ];
+  extraBuildInputs = [ ssreflect ];
```
Comment (vbgl) [inline pkgs/development/coq-modules/coquelicot/default.nix:20]: Library `ssreflect` must be propagated.

File: pkgs/development/coq-modules/coq-bits/default.nix (hunk 2 above)
Comment (ulrikstrid) [reply pkgs/development/coq-modules/coq-bits/default.nix:25]: Yes, I can't build the package without it. But I have now moved them to `extraBuildInputs` instead

File: pkgs/development/coq-modules/coquelicot/default.nix (hunk 3 above)
Comment (ulrikstrid) [reply pkgs/development/coq-modules/coquelicot/default.nix]: Reverted that change, thank you

**Comment by ulrikstrid** [comment]
Fixed the review comments from you @vbgl and `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` gives the same output as before so everything that  I touched seems to build.

**Review by Zimmi48 — APPROVED** [review]
The doc change LGTM.

**Review by vbgl — APPROVED** [review]
I must admit that I don’t fully understand `nativeBuildInputs` and `strictDeps`, but it looks OK now.
